src/mcp2515.cpp

```diff
@@ -650,7 +612,12 @@ MCP2515::ERROR MCP2515::sendMessage(const TXBn txbn, const struct can_frame *fra
 
   modifyRegister(txbuf->CTRL, TXB_TXREQ, TXB_TXREQ);
 
-  uint8_t ctrl = readRegister(txbuf->CTRL);
+  // wait for transmission to complete
+  uint8_t ctrl = 0x00;
+  do {
+    ctrl = readRegister(txbuf->CTRL);
+  } while (ctrl & TXB_TXREQ);
```

Comment:
Aha! I hadn't realize that I was the one who added this do-while loop... :)

I've noticed one instability with this. This while loop causes the bootloader to be indefinitely blocked on [the initial sendMessage() call](https://github.com/crycode-de/mcp-can-boot/blob/main/src/bootloader.cpp#L146) if the MCU is the only one on the bus. I think the MCP2515 is waiting for a device to ACK the frame, and therefore it never clear the TXREQ bit.

I might need to resolve this because that blocking issue gets annoying if you don't have any of devices on a bus during development/testing.

Reply:
I added a timeout to the loop.

I'll be honest, it took me while to iron this out. The MCP2515 datasheet has a flowchart that really helps explain the logic behind the transmit buffers. Like my comment says, the error bits are purely informational, they don't mean the MCP will stop trying to transmit the frame once an error condition is hit. It took me a while to realize that.

I can understand why the original developer probably didn't want to make this routine blocking. Since the MCP2515 can manage retransmissions on its own, it enables the user code to run off and do meaningful things with tha time. This bootloader doesn't have anything else meaningful to do, so it's okay for us to block and wait for the tx buffer to free itself up.